### PR TITLE
baggage: more efficient member validation

### PR DESCRIPTION
**Pull request by kevinburkesegment**
Instead of performing many, many comparisons for each character in a key and value, create an array of 128 bits and test for membership in that array, which is quicker.

Benchmarks on a Mac M1, ran 20 times.

```
name                             old time/op    new time/op    delta
New-8                               747ns ± 4%     741ns ± 3%    ~     (p=0.089 n=17+19)
NewMemberRaw-8                     11.4ns ± 1%    10.9ns ± 1%  -4.67%  (p=0.000 n=17+19)
Parse-8                             619ns ± 2%     603ns ± 1%  -2.54%  (p=0.000 n=18+18)
String-8                           1.12µs ± 4%    1.09µs ± 2%  -2.65%  (p=0.000 n=19+18)
ValueEscape/nothing_to_escape-8    6.93ns ± 1%    6.94ns ± 0%  +0.18%  (p=0.012 n=18+20)
ValueEscape/requires_escaping-8    26.3ns ± 2%    26.1ns ± 1%  -0.91%  (p=0.000 n=19+19)
ValueEscape/long_value-8            641ns ± 4%     595ns ± 1%  -7.14%  (p=0.000 n=18+17)
```

The approach is taken from the same approach in `encoding/json` in the standard library, which I also contributed.

### CODE REVIEW

**Comment by kevinburkesegment** [comment]
I don't think this PR needs a changelog entry.

**Review by XSAM — APPROVED** [review]
I am ok with this

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:634]: Have you compared of there will be the performance improvement if we change the type to `[utf8.RuneSelf]struct{}` and use the `_, ok := safeKeyCharset[s]` idiom?

**Comment by pellared** [comment]
> I don't think this PR needs a changelog entry.

We tend to add changelog entries when we improve performance.

**Comment by pellared** [comment]
> The approach is taken from the same approach in encoding/json in the standard library, which I also contributed.

Can you please provide a reference?

File: baggage/baggage.go
Comment (kevinburkesegment) [reply baggage/baggage.go:634]: Got a lot worse with a `map[byte]struct{}`. 

```
$ benchstat /tmp/main.txt /tmp/struct-variant.txt
name                             old time/op    new time/op    delta
New-8                               905ns ±11%    1136ns ± 8%   +25.46%  (p=0.000 n=19+19)
NewMemberRaw-8                     13.1ns ± 5%    43.4ns ±17%  +230.30%  (p=0.000 n=17+19)
Parse-8                             758ns ±10%    1594ns ± 6%  +110.14%  (p=0.000 n=20+16)
String-8                           1.33µs ±11%    4.52µs ±14%  +238.77%  (p=0.000 n=18+20)
ValueEscape/nothing_to_escape-8    7.97ns ± 4%   54.75ns ± 4%  +587.22%  (p=0.000 n=19+17)
ValueEscape/requires_escaping-8    31.1ns ± 4%    83.7ns ± 7%  +169.38%  (p=0.000 n=18+18)
ValueEscape/long_value-8            758ns ± 5%    5503ns ± 9%  +625.99%  (p=0.000 n=19+20)

name                             old alloc/op   new alloc/op   delta
New-8                                832B ± 0%      832B ± 0%      ~     (all equal)
NewMemberRaw-8                      0.00B          0.00B           ~     (all equal)
Parse-8                              832B ± 0%      832B ± 0%      ~     (all equal)
String-8                           1.06kB ± 0%    1.06kB ± 0%      ~     (all equal)
ValueEscape/nothing_to_escape-8     0.00B          0.00B           ~     (all equal)
ValueEscape/requires_escaping-8     16.0B ± 0%     16.0B ± 0%      ~     (all equal)
ValueEscape/long_value-8             576B ± 0%      576B ± 0%      ~     (all equal)

name                             old allocs/op  new allocs/op  delta
New-8                                16.0 ± 0%      16.0 ± 0%      ~     (all equal)
NewMemberRaw-8                       0.00           0.00           ~     (all equal)
Parse-8                              7.00 ± 0%      7.00 ± 0%      ~     (all equal)
String-8                             18.0 ± 0%      18.0 ± 0%      ~     (all equal)
ValueEscape/nothing_to_escape-8      0.00           0.00           ~     (all equal)
ValueEscape/requires_escaping-8      1.00 ± 0%      1.00 ± 0%      ~     (all equal)
ValueEscape/long_value-8             2.00 ± 0%      2.00 ± 0%      ~     (all equal)
```

File: baggage/baggage.go
Comment (pellared) [inline baggage/baggage.go:636]:
```suggestion
	'#': true,
	'$': true,
	'%': true,
	'&': true,
	'\'': true,
```

**Review by pellared — APPROVED** [review]

**Comment by pellared** [comment]
@open-telemetry/go-approvers waiting 1 day before merging.

**Review by dmathieu — APPROVED** [review]

**Pull request merged**
